src/sagas/dependency.saga.js

```diff
@@ -2,91 +2,171 @@
-    yield put(updateDependencyError(projectId, dependencyName));
+    yield call([console, console.error], 'Failed to install dependencies', err);
+    yield put(installDependenciesError(projectId, dependencies));
   }
 }
 
-/**
- * Trying to delete dependency, if success dispatching "finish" action,
- * if not - dispatching "error" action
- */
-export function* deleteDependency({
+export function* startUninstallingDependencies({
   projectId,
-  dependencyName,
+  dependencies,
 }: Action): Saga<void> {
   const projectPath = yield select(getPathForProjectId, projectId);
+
   try {
-    yield call(uninstallDependency, projectPath, dependencyName);
-    yield put(deleteDependencyFinish(projectId, dependencyName));
+    yield call(uninstallDependencies, projectPath, dependencies);
+    yield put(uninstallDependenciesFinish(projectId, dependencies));
   } catch (err) {
-    yield call([console, 'error'], 'Failed to delete dependency', err);
-    yield put(deleteDependencyError(projectId, dependencyName));
+    yield call(
+      [console, console.error],
+      'Failed to uninstall dependencies',
+      err
+    );
+    yield put(uninstallDependenciesError(projectId, dependencies));
   }
 }
 
-/**
- * Root dependencies saga, watching for "start" actions
- */
+export function* handleQueueActionCompleted({ projectId }: Action): Saga<void> {
+  yield put(startNextActionInQueue(projectId));
+}
+
+export function* handleNextActionInQueue({ projectId }: Action): Saga<void> {
+  const nextAction = yield select(getNextActionForProjectId, projectId);
+
+  // if the queue is empty, take no further action
+  if (!nextAction) return;
+
+  const actionCreator =
+    nextAction.action === 'install'
+      ? installDependenciesStart
+      : uninstallDependenciesStart;
+  yield put(actionCreator(projectId, nextAction.dependencies));
+}
+
+// Installs/uninstalls fail silently - the only notice of a failed action
+// visible to the user is either the dependency disappearing entirely or
+// having its status set back to `idle`.
+// TODO: display an error message outside of the console when a dependency
+// action fails
 export default function* rootSaga(): Saga<void> {
-  yield takeEvery(ADD_DEPENDENCY_START, addDependency);
-  yield takeEvery(UPDATE_DEPENDENCY_START, updateDependency);
-  yield takeEvery(DELETE_DEPENDENCY_START, deleteDependency);
+  yield takeEvery(ADD_DEPENDENCY, addDependency);
+  yield takeEvery(UPDATE_DEPENDENCY, updateDependency);
+  yield takeEvery(DELETE_DEPENDENCY, deleteDependency);
+  yield takeEvery(INSTALL_DEPENDENCIES_START, startInstallingDependencies);
+  yield takeEvery(UNINSTALL_DEPENDENCIES_START, startUninstallingDependencies);
+  yield takeEvery(
+    [
+      INSTALL_DEPENDENCIES_ERROR,
+      INSTALL_DEPENDENCIES_FINISH,
+      UNINSTALL_DEPENDENCIES_ERROR,
+      UNINSTALL_DEPENDENCIES_FINISH,
+    ],
+    handleQueueActionCompleted
+  );
+  yield takeEvery(START_NEXT_ACTION_IN_QUEUE, handleNextActionInQueue);
```

Comment:
Agreed, but then it conflicts with the action creator by the same name :(

I could follow the pattern we've adopted when fixing `no-shadow` and `import * as actions from '../actions'`, how about that?

Reply:
ohh, I see... right, actually it's probably best to avoid that confusion even if they're not imported.

How about we rename all of them to be `handleX`? So, `handleInstallDependenciesStart`, `handleAddDependency`, etc?